File: frontend/src/lib/api/location-groups.ts
import { apiRequest } from '@/lib/api/client';

export type CreateLocationGroupResponse = {
  group_code: string;
  group_key: string;
  created_at: string;
  storage: string;
  warning: string;
};

export type JoinLocationGroupInput = {
  client_id: string;
  display_name: string;
};

export type JoinLocationGroupResponse = JoinLocationGroupInput & {
  group_code: string;
  joined_at: string;
};

export type LocationUpdateInput = {
  latitude: number;
  longitude: number;
};

export type MemberLocation = {
  client_id: string;
  display_name: string;
  latitude: number;
  longitude: number;
  server_timestamp: string;
  stale: boolean;
};

export type LocationGroupSnapshot = {
  group_code: string;
  members: MemberLocation[];
  server_timestamp: string;
  stale_after_seconds: number;
  polling_interval_seconds: number;
};

export async function createLocationGroup(
  request: typeof apiRequest = apiRequest,
): Promise<CreateLocationGroupResponse> {
  return request<CreateLocationGroupResponse>('/api/v1/location-groups', {
    method: 'POST',
  });
}

export async function joinLocationGroup(
  groupCode: string,
  groupKey: string,
  input: JoinLocationGroupInput,
  request: typeof apiRequest = apiRequest,
): Promise<JoinLocationGroupResponse> {
  return request<JoinLocationGroupResponse>(
    '/api/v1/location-groups/' + encodeURIComponent(groupCode) + '/join',
    {
      method: 'POST',
      headers: { 'X-Group-Key': groupKey },
      body: input,
    },
  );
}

export async function updateLocationGroupMember(
  groupCode: string,
  groupKey: string,
  clientId: string,
  input: LocationUpdateInput,
  request: typeof apiRequest = apiRequest,
): Promise<MemberLocation> {
  return request<MemberLocation>(
    '/api/v1/location-groups/' +
      encodeURIComponent(groupCode) +
      '/members/' +
      encodeURIComponent(clientId) +
      '/location',
    {
      method: 'PUT',
      headers: { 'X-Group-Key': groupKey },
      body: input,
    },
  );
}

export async function fetchLocationGroupSnapshot(
  groupCode: string,
  groupKey: string,
  request: typeof apiRequest = apiRequest,
): Promise<LocationGroupSnapshot> {
  return request<LocationGroupSnapshot>(
    '/api/v1/location-groups/' +
      encodeURIComponent(groupCode) +
      '/locations',
    {
      headers: { 'X-Group-Key': groupKey },
    },
  );
}

File: backend/code/src/models/location_groups.py
from datetime import datetime

from pydantic import BaseModel, Field


class CreateLocationGroupResponse(BaseModel):
    group_code: str = Field(min_length=6, max_length=6)
    group_key: str
    created_at: datetime
    storage: str = "memory_only"
    warning: str


class JoinLocationGroupRequest(BaseModel):
    client_id: str = Field(min_length=3, max_length=80)
    display_name: str = Field(min_length=1, max_length=80)


class JoinLocationGroupResponse(BaseModel):
    group_code: str
    client_id: str
    display_name: str
    joined_at: datetime


class LocationUpdateRequest(BaseModel):
    latitude: float = Field(ge=-90, le=90)
    longitude: float = Field(ge=-180, le=180)


class MemberLocation(BaseModel):
    client_id: str
    display_name: str
    latitude: float
    longitude: float
    server_timestamp: datetime
    stale: bool


class LocationGroupSnapshot(BaseModel):
    group_code: str
    members: list[MemberLocation]
    server_timestamp: datetime
    stale_after_seconds: int
    polling_interval_seconds: int
